### ISSUE TITLE
Compiling mx Modules

### ISSUE THREAD

**Issue by doug777** [prominic/Moonshine-IDE#626]
- [ ] Review the documentation: https://github.com/prominic/Moonshine-IDE
- [ ] Search for existing issues: https://github.com/prominic/Moonshine-IDE/issues
- [ ] Use the latest stable version at: http://moonshine-ide.com
- [ ] Let us know how to reproduce the issue. Include a code sample, or share a project that reproduces the issue

## Environment
Win10 Pro v1903

## Description
I have recently transferred into Moonshine an ancient AIR desktop project written entirely in mx and AS3 that uses Modules extensively.

Is there a way in Moonshine to compile the modules and then create the AIR package?

In FlexBuilder there is a file called .actionScriptProperties with a node called -modules- where each module is listed with its path. When you run the compiler it uses the data in this node to create a mirror of the project folders containing each module now an swf, in place of the original mxml module.

Is there any way to duplicate this compile function in Moonshine and then create the AIR file?

Doug

## Steps to Reproduce
Application builds and runs okay but of course it cannot load any module.

## Expected Behavior
Hoped it would compile everything like FlexBuilder

**Comment by doug777** [comment]
If I compile each module separately and place the swfs into the correct directory structure, when I debug, the project builds correctly but when it starts to load the first module I get an error -

Illegal override of FlexModuleFactory in mx.core.FlexModuleFactory

This class contains a function allowDomain() which perhaps is the problem as it is commented - this function is overridden by the compiler. 

Is it possible that the Moonshine compiler doesn't do this? Is there some way to get the compiler to do this by adding something to the compiler options?

Any help would be appreciated as this project contains over 50 modules and changing all that to use a different method would be a big task.

Doug

**Comment by piotrzarzycki21** [comment]
I'm not sure how to help you @doug777 . I can definitely encourage you to try [dev version](https://moonshine-ide.com/downloads/releases/dev/windows/) of Moonshine and see how it's working.

**Comment by doug777** [comment]
I've got the latest dev version now but there's no difference.

I have found a way to use FB3 to compile main app and all the modules and then use AIR32's adt to create the AIR package. When I run the air file it creates, it works correctly and there is no FlexModuleFactory error.

But I desperately need to debug in Moonshine.

I think it's likely that the error is caused by the Moonshine compiler missing something when it handles ModuleLoader so that something sends null to FlexModuleFactory when ModuleLoader is called.

I have added to bin-debug all the necessary files to create a correctly working app if packaged by adt. But I guess Moonshine uses Ant to do the compiling/packaging and is missing something that modules need?

Is there any way to extend the debugger so I can set breakpoints in the sdk classes to see what's causing the error?

Doug

**Comment by piotrzarzycki21** [comment]
@joshtynjala do you have any idea what we may miss in debugger?

**Comment by joshtynjala** [comment]
As far as I know, there's no way to launch a debug session in Moonshine without also compiling first, so I don't think it will be possible to debug Doug's app in Moonshine, even if he can use Flex Builder 3 to compile it.

With that in mind, the first issue that we would need to resolve is how to properly compile Flex modules in Moonshine. I'm not super familiar with how Moonshine calls the Flex SDK compiler. However, I just looked at the settings available in a Flex project, and I don't see any fields where you can specify which files are modules. That seems like pretty good evidence that Moonshine doesn't know how to compile Flex modules at all.

Basically, I don't think that this a debugger issue. It's how the app is compiled.

**Comment by doug777** [comment]
Yes I'm sure Josh is correct, but if you add the compiled modules to the bin-debug folder or the bin-release folder, they remain untouched by the compiler which only overwrites MyApp.mxml and it should work, but it causes the FlexModuleFactory error both in Debug AND Run. So it's definitely not the debugger. It is the compiler that has a problem.

*** I do not need this since I can do it manually, but -

To compile everything in Moonshine it would be necessary to add a Modules option to the Paths settings. This would need to enable you to add each module and set source path and whether to optimize for a specific applicaton or not (if it is used by more than one application) with This App as the default. Once added a typical setting would look like this -

Source path - src\com\mycompany\modules\MyModule.mxml
Output SWF  - com\mycompany\modules\MyModule.swf
Optimize for- MyApp.mxml

Then the compiler would have to compile each module listed and place the swf in the output path specified. So bin-debug would now contain MyApp.xml and a com folder with all the compiled modules.

When you call a module with ModuleLoader the code looks something like this

mx:ModuleLoader id="mlLogin" url="app:/com/mycompany/modules/Login.swf" width="100%" height="100%" applicationDomain="{currDomain}"/
where currDomain = ApplicationDomain.currentDomain

So it is not unreasonable to assume that if you separately compile the modules and create the com folder and place it in bin-debug/bin-release it would still work since it will find the compiled module at the url location.

*** What I need is this -

It's difficult to know exactly what happens in FlexModuleFactory.as since I can't debug it, but it looks like the Moonshine compiler contains the override code for FlexModuleFactory, but then doesn't pass to it ModuleLoader's applicationDomain parameter correctly, probably it's passing null and hence the error -

VerifyError: Error #1053: Illegal override of FlexModuleFactory in mx.core.FlexModuleFactory. 

This is the relevant code in FlexModuleFactory -

    /**
     *  Calls Security.allowDomain() for the SWF associated with this FlexModuleFactory.
     * 
     */  
    public function allowDomain(... domains):void
    {
        // Overridden by compiler generated code.
    }
    
So I think my project could be run and debugged succesfully if the compiler correctly set the argument.

Security is in the flash.system package, which I can't see at all, so no way to know if this is everything needed, but the details are here -

https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/system/Security.html#allowDomain()

*** Okay I realize this is a big ask and no guarantee that this is the solution, but please let me know if there is anything you can do. I can in fact work around this by using FB3 and adt, but I would prefer to use Moonshine if possible.

Doug

**Comment by piotrzarzycki21** [comment]
@doug777 Could you try latest release or Nightly build ?
